test/e2e/tests/evidence_test.go

```diff
@@ -0,0 +1,98 @@
+package e2e_test
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	cs "github.com/tendermint/tendermint/test/e2e/maverick/consensus"
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+	"github.com/tendermint/tendermint/types"
+)
+
+// assert that all nodes that have blocks during the height (or height + 1) of a misbehavior has evidence
+// for that misbehavior
+func TestEvidence_Misbehavior(t *testing.T) {
```

Comment:
I think this test is unnecessarily complicated. I would do something like this:

```go
func TestEvidence_Misbehavior(t *testing.T) {
	blocks := fetchBlockChain(t)
	testNode(t, func(t *testing.T, node e2e.Node) {
		for _, block := range blocks {
			// Find any evidence blaming this node in this block
			var nodeEvidence types.Evidence
			for _, evidence := range block.Evidence.Evidence {
				switch evidence := evidence.(type) {
				case *types.DuplicateVoteEvidence:
					if bytes.Equal(evidence.VoteA.ValidatorAddress, node.Key.PubKey().Address()) {
						nodeEvidence = evidence
					}
				default:
					t.Fatalf("unexpected evidence type %T", evidence)
				}
			}

			// Check that evidence was as expected (evidence is submitted in following height)
			misbehavior, ok := node.Misbehaviors[block.Height-1]
			if !ok {
				require.Nil(t, nodeEvidence, "found unexpected evidence %v in height %v",
					nodeEvidence, block.Height)
				continue
			}
			require.NotNil(t, nodeEvidence, "no evidence found for misbehavior %v in height %v",
				misbehavior, block.Height)

			switch misbehavior {
			case "double-prevote":
				require.IsType(t, &types.DuplicateVoteEvidence{}, nodeEvidence, "unexpected evidence type")
			default:
				t.Fatalf("unknown misbehavior %v", misbehavior)
			}
		}
	})
}
```

Reply:
Okay I think I wanted to check that all nodes that were participating at that height each had a copy of the evidence but perhaps this is redundant

Reply:
Ah, I see. Yes, we have a separate test `TestBlock_Header` which checks that all blocks are identical across all nodes, so we can assume that `fetchBlockChain()` data is valid everywhere. Otherwise I'd agree with doing an exhaustive search.